sdks/python/apache_beam/io/filesystem.py

```diff
@@ -0,0 +1,439 @@
+    if not self._decompressor:
+      raise ValueError('decompressor not initialized')
+
+    io = cStringIO.StringIO()
+    while True:
+      # Ensure that the internal buffer has at least half the read_size. Going
+      # with half the _read_size (as opposed to a full _read_size) to ensure
+      # that actual fetches are more evenly spread out, as opposed to having 2
+      # consecutive reads at the beginning of a read.
+      self._fetch_to_internal_buffer(self._read_size / 2)
+      line = self._read_from_internal_buffer(
+          lambda: self._read_buffer.readline())
+      io.write(line)
+      if line.endswith('\n') or not line:
+        break  # Newline or EOF reached.
+
+    return io.getvalue()
+
+  def closed(self):
+    return not self._file or self._file.closed()
+
+  def close(self):
+    if self.readable():
+      self._read_buffer.close()
+
+    if self.writeable():
+      self._file.write(self._compressor.flush())
+
+    self._file.close()
+
+  def flush(self):
+    if self.writeable():
+      self._file.write(self._compressor.flush())
+    self._file.flush()
+
+  @property
+  def seekable(self):
+    # TODO: Add support for seeking to a file position.
+    return False
+
+  def tell(self):
+    """Returns current position in uncompressed file."""
+    return self._uncompressed_position
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, exception_type, exception_value, traceback):
+    self.close()
+
+
+class FileMetadata(object):
+  """Metadata about a file path that is the output of FileSystem.match
+  """
+  def __init__(self, path, size_in_bytes):
+    assert isinstance(path, basestring) and path, "Path should be a string"
+    assert isinstance(size_in_bytes, int) and size_in_bytes >= 0, \
+        "Size of bytes should be greater than equal to zero"
+    self.path = path
+    self.size_in_bytes = size_in_bytes
+
+  def __eq__(self, other):
+    """Note: This is only used in tests where we verify that mock objects match.
+    """
+    return (isinstance(other, FileMetadata) and
+            self.path == other.path and
+            self.size_in_bytes == other.size_in_bytes)
+
+  def __hash__(self):
+    return hash((self.path, self.size_in_bytes))
+
+  def __ne__(self, other):
+    return not self.__eq__(other)
+
+  def __repr__(self):
+    return 'FileMetadata(%s, %s)' % (self.path, self.size_in_bytes)
+
+
+class MatchResult(object):
+  """Result from the ``FileSystem`` match operation which contains the list
+   of matched FileMetadata.
+  """
+  def __init__(self, pattern, metadata_list):
+    self.metadata_list = metadata_list
+    self.pattern = pattern
+
+
+class BeamIOError(IOError):
+  def __init__(self, msg, exception_details=None):
+    """Class representing the errors thrown in the batch file operations.
+    Args:
+      msg: Message string for the exception thrown
+      exception_details: Map of indivisual input to exception for failed
```

Comment:
Please expand the doc here to include more details. For example,

(1) is 'exception_details' required or optional
(2) what could the state of the file-system be if 'exception_details' is specified/unspecified.
(3) How should the callers use 'exception_details'

My current understanding is that 'exception_details' is optional and will simplify callers if specified (callers can only be simplified if they choose to fail -- or ignore in some cases -- when 'exception_details' is not available).